src/reconcile/rec_visibility.c

```diff
@@ -552,20 +552,29 @@ __rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_UPDATE *first_upd
          * check is not required for history store updates as they are implicitly committed. As
          * prepared transaction IDs are globally visible, need to check the update state as well.
          *
-         * If an earlier reconciliation chose this update (it is marked as being destined for the
-         * data store), we should select it regardless of visibility if we haven't already selected
-         * one. This is important as it is never ok to shift the on-disk value backwards in the
-         * update chain.
+         * There are several cases we should select the update irrespective of visibility:
          *
-         * Also, if an earlier reconciliation performed an update-restore eviction and this update
-         * was restored from disk, or, we rolled back a prepared transaction and restored an update
-         * from the history store, we can select this update irrespective of visibility. This
-         * scenario can happen if the current reconciliation has a limited visibility of updates
-         * compared to one of the previous reconciliations.
+         * 1. A previous reconciliation selected this update as writing anything that is older
+         * undoes the previous work.
+         *
+         * 2. The update is restored from the disk image as writing anything that is older undoes
+         * the previous work.
+         *
+         * 3. An earlier reconciliation performed an update-restore eviction and this update was
+         * restored from disk.
+         *
+         * 4. We rolled back a prepared transaction and restored an update from the history store.
+         *
+         * 5. We rolled back a prepared transaction and aim to delete the following update from the
+         * history store.
+         *
+         * These scenarios can happen if the current reconciliation has a limited visibility of
+         * updates compared to one of the previous reconciliations. This is important as it is never
+         * ok to undo the work of the previous reconciliations.
          */
         if (!F_ISSET(upd,
               WT_UPDATE_DS | WT_UPDATE_PREPARE_RESTORED_FROM_DS | WT_UPDATE_RESTORED_FROM_DS |
-                WT_UPDATE_RESTORED_FROM_HS) &&
+                WT_UPDATE_RESTORED_FROM_HS | WT_UPDATE_TO_DELETE_FROM_HS) &&
```

Comment:
The change makes sense. Do you think it would be worth creating a new define with these flags grouped together near their definition in the header?
Something like WT_UPDATE_ON_DISK? It feels as though that might make it less likely to be missed if another flag is added in the future.
Or is this a case where we thought about it and made the wrong decision?

Reply:
Yes, that's a good point. It is a mistake not adding the flag here. I'll create a new define.